oommf: adding a new package

fangohr opened the pull request:
Tested to work on Linux (https://github.com/fangohr/oommf-in-spack)

fangohr committed:
oommf: adding a new package

Tested to work on Linux (https://github.com/fangohr/oommf-in-spack)

fangohr committed:
oommf: remove f-strings in response to vermin test

Failed this before:
https://github.com/spack/spack/runs/3998910371?check_suite_focus=true#step:6:12

bernhardkaindl commented on var/spack/repos/builtin/packages/oommf/package.py:
```diff
@@ -0,0 +1,330 @@
+    version(
+        "20a3_20210930",
+        sha256="880242afdf4c84de7f2a3c42ab0ad8c354028a7d2d3c3160980cf3e08e285691",
+        preferred=True,
```
```suggestion
```
The version which is defined first is automatically the default version.

If there is no special reason (like if new versions should not be the preferred versions)
you don't specify `preferred=True`, and then you don't have to move it up to the new version you just added.

This would also come in handy when https://github.com/spack/spack/pull/24532 is merged to make adding new versions even easier. Then, we would not want maintainers to to use `preferred=True` for easy recipe updates, unless there is a specific reason for it.

(already commented in your repo: Added here as suggestion in case you don't see that comment).
Of course, you can decline it.

bernhardkaindl reviewed (changes requested):
Building `20a3_20210930` with `spack install --test=root -v oommf %clang@13.0.0`, I got this error:
```cxx
spack-src/oommf/pkg/oc/oc.cc: In function 'int Oc_Init(Tcl_Interp*)':
spack-src/oommf/pkg/oc/oc.cc:1513:6: error: 'OC_SYSTEM_TYPE' was not declared in this scope
   if(OC_SYSTEM_TYPE == OC_UNIX) {
      ^~~~~~~~~~~~~~
```
The build with `%clang@11.0.1` worked, but it looks like at least some part of the build and the tests isn't using spack's compiler wrapper:
```
g++ -c -DNDEBUG -Wno-non-template-friend -Wno-misleading-indentation -Ofast -ffast-math -std=c++11 -frename-registers -fstrict-aliasing -fweb -fomit-frame-pointer -momit-leaf-frame-pointer -I/home/kaindlb/home/spack/.stage/spack-stage-oommf-20a3_20210930-
```  
```
Testing oommf.tcl +platform
==> [2021-10-26-18:12:45.724271] '/home/kaindlb/home/spack/.tree1/linux-ubuntu18.04-zen2/clang-11.1.0/tcl-8.6.11-kzs7cut7mjahjd26ussdimmdabak2lbc/bin/tclsh' '/home/kaindlb/home/spack/.tree1/linux-ubuntu18.04-zen2/clang-11.1.0/oommf-20a3_20210930-nfjevztqsoocc6d75owzvgkuuxnrmpus/bin/oommf.tcl' '+platform'
OOMMF release 2.0a3
Platform Name:          linux-x86_64
Tcl name for OS:        Linux 4.15.0-117-generic
C++ compiler:           /usr/bin/g++
 Version string:         g++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0 / x86_64-linux-gnu
```
Here is an analysis showing it called /usr/bin/g++ of the Ubuntu host, when running `spack install --test=root -v oommf %clang@12.0.1`. To check, I suspended `spack` with Ctrl-Z and ran `ps aux|grep ++` to be able to check the suspended compiler processes:
```
...  44256  0.0  0.0   5700   840 pts/3    T    18:29   0:00 g++ -c -DNDEBUG... spack-src/oommf/app/oxs/local/oommf-mel/yy_transformstagemel.cc
/usr/lib/gcc/x86_64-linux-gnu/7/cc1plus ...
...  44262  0.0  0.0  13332  1112 pts/5    S+   18:29   0:00 grep g++
$ ls -l /proc/44256/exe
lrwxrwxrwx 1 kaindlb kaindlb 0 Oct 26 18:30 /proc/44256/exe -> /usr/bin/x86_64-linux-gnu-g++-7
```
Whereas a build with %gcc@11.2.0 seems to use it (and with the spack compiler wrapper, providing the optimisation flags for the CPU):
```py
.../gcc-11.2.0-rykaqzhni45tv2jmg7z5iawxmvxatwkp/bin/g++ -march=znver2 -mtune=znver2
```

bernhardkaindl commented on var/spack/repos/builtin/packages/oommf/package.py:
```diff
@@ -0,0 +1,330 @@
+    def setup_build_environment(self, env):
+        # Seems we don't need to do anything here
+        pass
+
```
```suggestion
```
When you don't use it, remove it.

bernhardkaindl commented on var/spack/repos/builtin/packages/oommf/package.py:
```diff
@@ -0,0 +1,330 @@
+        tclsh = Executable(join_path(spec["tcl"].prefix.bin, "tclsh"))
+        # where is oommf.tcl?
+        oommf_tcl_path = join_path(spec.prefix.bin, "oommf.tcl")
+        # put the command together and execute
```
The comments here are redundant, the variable assignments speak for themselfes.
Another tip: Comment they **why** not the **what**.

For larger blocks of code which are hard to read, you might consider turning them into class methods, where the method name and the method call are basically replacing the comment and make the code and the function where it's used easier to read (if you can, just a suggestion).

Also, you get these specific values three times:

What about turning these into an `@property` of the class so you can just use it like a global variable directly? (See other recipes for example, use `git grep -A3 @property var/spack/repos/builtin` to see the example of `var/spack/repos/builtin/packages/bohrium/package.py`)

fangohr replied:
Thank you. The properties are implemented, and indeed code duplication is much reduced.

bernhardkaindl commented on var/spack/repos/builtin/packages/oommf/package.py:
```diff
@@ -0,0 +1,330 @@
+            output=str.split,
+            error=str.split,
+            env=test_env,
+        )
```
For me, this looks more readable:
```suggestion
        output = tclsh(oommf_tcl_path, "boxsi", "+fg", task, "-exitondone", "1",
            output=str.split, error=str.split, env=test_env)
```

fangohr replied:
I ran this through black. If you don't mind, I'll leave it as is so that I can keep doing that for future updates.

bernhardkaindl commented:
@fangohr Hi, in case you need OOMMF in spack ASAP, since it looks mostly ok when using gcc, I'd not be opposed to merge it with a `conflicts('%clang')`, including a comment which explains why it is there.

Mostly, I wanted to make you aware of the things that I saw with the current code.

bernhardkaindl reviewed (changes requested):
From the successful build of `20a3_20210930` using `--test=root` with gcc-11, three test programs are still running, long after install and test finished:
```bash
ps `pgrep -f gcc-11`|sed -n s/.*oo/oo/p
oommf/pkg/net/threads/host.tcl -tk 0 15136 39151
oommf/pkg/net/threads/account.tcl -tk 0 0 34939
oommf/app/mmarchive/mmarchive.tcl -tk 0
```

fangohr commented:
@bernhardkaindl Excellent, educational and constructive feedback - many thanks!

fangohr committed:
Apply suggestions from code review

Remove unnecessary code.

Co-authored-by: Bernhard Kaindl <43588962+bernhardkaindl@users.noreply.github.com>

fangohr committed:
conflict for clang, point to MR discussion

Also created a todo for this in the repo we used to prepare the package.py
file: https://github.com/fangohr/oommf-in-spack/issues/31

fangohr committed:
Reduce duplication

- through introduction of properties
- introduction of _check_install_oommf_command()

(Based on suggestions from @bernhardkaindl and @lang-m)

fangohr committed:
typo fix

fangohr committed:
most recent OOMMF 20a3 has changed its output for +version

so we need to check for a string that is in old and current version.

(@lang-m spotted this)

fangohr committed:
Stop all OOMMF processes

at the end of execution of the command.

Addresses https://github.com/spack/spack/pull/26933#pullrequestreview-789840519

fangohr committed:
reformatting (black)

fangohr commented:
> @fangohr Hi, in case you need OOMMF in spack ASAP, since it looks mostly ok when using gcc, I'd not be opposed to merge it with a `conflicts('%clang')`, including a comment which explains why it is there.

I like that suggestion (is implemented).

fangohr commented:
@bernhardkaindl  - I believe all feedback from the first review is addressed. Thank you for the guidance. 
Have you got further feedback and suggestions for improvements?

PS One CI test fails (https://gitlab.spack.io/spack/spack/-/pipelines/77463) which looks to be independent from this MR.

bernhardkaindl reviewed (approved):
Looks great and builds fine with gcc-10 and -11. But the CI pipeline has to pass for the merge.

bernhardkaindl commented:
Closing and reopening as `spackbot run-pipeline` didn't react.

fangohr commented:
@bernhardkaindl - what can we do to rerun the CI (hoping this will fix it) or get help?

fangohr commented:
> @bernhardkaindl - what can we do to rerun the CI (hoping this will fix it) or get help?

I don't know what changed, but I think you could press the Merge button now. If so, can you do it? Many thanks!
